src/main_segment/calc.c

```diff
@@ -54,31 +186,147 @@ INCLUDE_ASM("asm/nonmatchings/main_segment/calc", func_8007C9C8);
+ * outY, outZ)`.
+ *
+ * (This is equivalent to multiplying by the inverse of `m` if the linear part of `m` is orthogonal.)
+ */
+void func_8007CC68(MtxF *m, f32 inX, f32 inY, f32 inZ, f32 *outX, f32 *outY, f32 *outZ) {
+    f32 vec[3];
+    f32 tempX = inX - m->mf[3][0];
+    f32 tempY = inY - m->mf[3][1];
+    f32 tempZ = inZ - m->mf[3][2];
+    s32 i;
+
+    for (i = 0; i < 3; i++) {
+        vec[i] = (m->mf[i][0] * tempX) + (m->mf[i][1] * tempY) + (m->mf[i][2] * tempZ);
+    }
+
+    *outX = vec[0];
+    *outY = vec[1];
+    *outZ = vec[2];
+}
 
-INCLUDE_ASM("asm/nonmatchings/main_segment/calc", func_8007CC68);
+/**
+ * Multiply the vector `(inX, inY, inZ)` by the linear part of the matrix `m` and write to `(outX, outY, outZ)`.
+ *
+ * (This is equivalent to multiplying by the inverse of the linear part of `m`, if the linear part of `m` is orthogonal.)
+ */
+void func_8007CCFC(MtxF *m, f32 inX, f32 inY, f32 inZ, f32 *outX, f32 *outY, f32 *outZ) {
+    f32 vec[3];
+    s32 i;
 
-INCLUDE_ASM("asm/nonmatchings/main_segment/calc", func_8007CCFC);
+    for (i = 0; i < 3; i++) {
+        vec[i] = (m->mf[i][0] * inX) + (m->mf[i][1] * inY) + (m->mf[i][2] * inZ);
+    }
+    *outX = vec[0];
+    *outY = vec[1];
+    *outZ = vec[2];
+}
 
 INCLUDE_ASM("asm/nonmatchings/main_segment/calc", func_8007CD78);
 
-INCLUDE_ASM("asm/nonmatchings/main_segment/calc", func_8007CFB4);
+/**
+ * Normalise a vector consisting of three passed floats in-place. Returns the length of the original vector.
+ */
+f32 func_8007CFB4(f32 *x, f32 *y, f32 *z) {
+    f32 tempX = *x;
+    f32 tempY = *y;
+    f32 tempZ = *z;
+    f32 length = sqrtf((tempX * tempX) + (tempY * tempY) + (tempZ * tempZ));
+
+    if (length != 0.0f) {
+        tempX /= length;
+        tempY /= length;
+        tempZ /= length;
+    }
+    *x = tempX;
+    *y = tempY;
+    *z = tempZ;
+    return length;
+}
+
+void randomize00(void) {
+    s32 oldValue1;
+    s32 oldValue2;
+    s32 newValue;
+    s32 i;
+
+    for (i = 0; i < 24; i++) {
+        oldValue1 = randomtable[i];
+        oldValue2 = randomtable[i + 31];
+        newValue = oldValue1 - oldValue2;
+        if (newValue < 0) {
+            newValue += 0x10000;
+        }
+        randomtable[i] = newValue;
+    }
 
-INCLUDE_ASM("asm/nonmatchings/main_segment/calc", randomize00);
+    for (i = 24; i < 55; i++) {
+        oldValue1 = randomtable[i];
+        oldValue2 = randomtable[i - 24];
+        newValue = oldValue1 - oldValue2;
+        if (newValue < 0) {
+            newValue += 0x10000;
+        }
+        randomtable[i] = newValue;
+    }
+}
 
-INCLUDE_ASM("asm/nonmatchings/main_segment/calc", randomseed);
+void randomseed(s32 seed) {
+    s32 value;
+    s32 i;
+    s32 index;
+
+    randomtable[54] = seed;
+    for (value = 1, i = 1; i < 55; i++) {
+        index = ((21 * i) % 55) - 1;
```

Comment:
Could you use `ARRAY_COUNT` here?

Reply:
Given that there's still blank space after the array, I'm still not convinced of the size.

Reply:
According to Dr_MARIO.plf, the size is right:
```
   Num:    Value  Size Type    Bind   Vis      Ndx Name
   437: 0001a0c4   220 OBJECT  LOCAL  DEFAULT    6 randomtable
```
(remember readelf uses dec for size)